_posts/outbox-what-and-why.md

```diff
@@ -0,0 +1,155 @@
+
+All that means we could be in a situation where an endpoint successfully stores data and messages (the intent to send them) in the storage but fails to deliver them to the queuing system. Or it could dispatch them but fail at marking messages as delivered.
+
+A temptation could be to use a background task. The endpoint could run a task on a background thread to check the storage for unsent messages and deliver them. However, more is needed to solve the problem. We cannot have a transaction that spans the two resources at play. It also introduces a new undesirable side effect: if the endpoint is scaled horizontally, multiple background threads will look at the same storage for undispatched messages. All of them will find messages to dispatch. To coordinate, they need a distributed transaction. Otherwise, we'd have to accept that they might send multiple copies of the same message.
+
+## The incoming message is our best friend
+
+It's in that context that an incoming message is beneficial. We can use incoming messages as resource coordinators. Why is that? Primarily because message identifiers are stable over time. If an endpoint retries the same message multiple times, it'll have the same message id. If the queuing infrastructure is "at least once," it could deliver multiple copies of the same message. All copies will have the same message-id.
+
+With an incoming message when the endpoint receives it, we could perform the following steps:
+
+1. Start a local transaction and store the message id in the outbox storage.
+  * The storage has a unique index on the message identifier.
+2. Handle the message from the business logic perspective, for example:
+  * Store some business data enlisting in the same local transaction
+  * Send a couple of messages (remember that these are dispatch intentions stored in the same storage enlisting in the same local transaction)
+3. Commit the local transaction making sure that:
+  * Data are safely stored.
+  * Messages we want to send are in the outbox storage.
+  * The incoming message id is in the database too.
+4. Send messages we want to send and mark them as sent (outside the original local transaction).
+5. Acknowledge the incoming message, deleting it from the queue.
+
+If anything between 1 and 3 fails, the local transaction gets rolled back, no changes are applied, and the endpoint will retry the incoming message.
+
+If what fails are either steps 4 or 5, the endpoint picks up the message again. The outbox already contains the message identifier, stored at the first step and committed at step three. The outbox treats it as an "already processed message for which something went wrong." In this mode, the endpoint skips processing the message and jumps to step 4, sending outgoing messages, marking them as sent, and finally acknowledging the incoming message.
```

Comment:
>  the endpoint skips processing the message

above you say 

> Handle the message from the business logic perspective

maybe it would be better to rephrase this to skip executing the business logic or something similar

Reply:
```suggestion
If what fails are either steps 4 or 5, the endpoint picks up the message again. The outbox already contains the message identifier, stored at the first step and committed at step three. The outbox treats it as an "already processed message for which something went wrong." In this mode, the endpoint skips invoking the users' business logic and jumps to step 4, sending outgoing messages, marking them as sent, and finally acknowledging the incoming message.
```

Thanks, Daniel.